x/collection/msgs.go

```diff
@@ -0,0 +1,515 @@
+	}
+
+	return nil
+}
+
+func validateID(id string, reg *regexp.Regexp) error {
+	if !reg.MatchString(id) {
+		return sdkerrors.ErrInvalidRequest.Wrapf("invalid id: %s", id)
+	}
+	return nil
+}
+
+func validateName(name string) error {
+	if err := validateStringSize(name, nameLengthLimit, "name"); err != nil {
+		return ErrInvalidNameLength.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+func validateURI(uri string) error {
+	if err := validateStringSize(uri, uriLengthLimit, "uri"); err != nil {
+		return ErrInvalidBaseImgURILength.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+func validateMeta(meta string) error {
+	if err := validateStringSize(meta, metaLengthLimit, "meta"); err != nil {
+		return ErrInvalidMetaLength.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+func validateStringSize(str string, limit int, name string) error {
+	if length := utf8.RuneCountInString(str); length > limit {
+		return sdkerrors.ErrInvalidRequest.Wrapf("%s cannot exceed %d in length: current %d", name, limit, length)
+	}
+	return nil
+}
+
+func validateLegacyPermission(permission string) error {
+	return ValidatePermission(Permission(LegacyPermissionFromString(permission)))
+}
+
+func ValidatePermission(permission Permission) error {
+	if p := Permission_value[Permission_name[int32(permission)]]; p == 0 {
+		return ErrInvalidPermission.Wrap(permission.String())
+	}
+	return nil
+}
+
+func validateContractChange(change Attribute) error {
+	validators := map[string]func(string) error{
+		AttributeKeyName.String():       validateName,
+		AttributeKeyBaseImgURI.String(): validateURI,
+		AttributeKeyMeta.String():       validateMeta,
+		AttributeKeyURI.String():        validateURI,
+	}
+
+	return validateChange(change, validators)
+}
+
+func validateTokenClassChange(change Attribute) error {
+	validators := map[string]func(string) error{
+		AttributeKeyName.String(): validateName,
+		AttributeKeyMeta.String(): validateMeta,
+	}
+
+	return validateChange(change, validators)
+}
+
+func validateChange(change Attribute, validators map[string]func(string) error) error {
+	validator, ok := validators[change.Key]
+	if !ok {
+		return ErrInvalidChangesField.Wrapf("invalid field: %s", change.Key)
+	}
+	return validator(change.Value)
+}
+
+// ValidateBasic implements Msg.
+func (m MsgSendNFT) ValidateBasic() error {
+	if err := ValidateContractID(m.ContractId); err != nil {
+		return err
+	}
+
+	if _, err := sdk.AccAddressFromBech32(m.From); err != nil {
+		return sdkerrors.ErrInvalidAddress.Wrapf("invalid from address: %s", m.From)
+	}
+	if _, err := sdk.AccAddressFromBech32(m.To); err != nil {
+		return sdkerrors.ErrInvalidAddress.Wrapf("invalid to address: %s", m.To)
+	}
+
+	if len(m.TokenIds) == 0 {
+		return ErrEmptyField.Wrap("token ids cannot be empty")
+	}
+	for _, id := range m.TokenIds {
+		if err := ValidateTokenID(id); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ValidateBasic implements Msg.
+func (m MsgOperatorSendNFT) ValidateBasic() error {
```

Comment:
It's good time to make messages' `ValidateBasic` method from value receiver to pointer receiver to follow `In general, all methods on a given type should have either value or pointer receivers, but not a mixture of both. (We'll see why over the next few pages.)`( https://go.dev/tour/methods/8)

```suggestion
func (m MsgOperatorSendNFT) ValidateBasic() error {
```
```suggestion
func (m *MsgOperatorSendNFT) ValidateBasic() error {
```